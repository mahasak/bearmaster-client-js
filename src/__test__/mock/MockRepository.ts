import { Storage } from '../../storage';
import Repository, {
    RepositoryOptions
} from '../../repository';
const appName = 'foo';
const instanceId = 'bar';
import { MockStorage } from './MockStorage'
import { IExperiment } from '../../interfaces';
import { buildToggle } from '../client/buildToggle'

export class MockRepository extends Repository {
    private toggleFunction: Function;
    constructor({
        backupPath,
        url,
        appName,
        instanceId,
        refreshInterval,
        StorageImpl = MockStorage,
        timeout,
        headers,
        customHeadersFunction,
    }: RepositoryOptions, toggleFunction: Function) {
        super({
            backupPath,
            url,
            appName,
            instanceId,
            refreshInterval,
            StorageImpl,
            timeout,
            headers,
            customHeadersFunction,
        });
        this.toggleFunction = toggleFunction;
    }


    async fetch() { }


    getToggle(name: string): IExperiment {
        return this.toggleFunction();
    }

    getToggles(): IExperiment[] {
        return this.toggleFunction();
    }

    getETag(): string {
        return this.etag ?? '';
    }

    setEtag(etag: string): void {
        this.etag = etag;
    }
}