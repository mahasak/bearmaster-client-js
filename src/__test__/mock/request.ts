export const request = {
    get: jest.fn( ()=> {
        console.log(
            'Invoked'
        );
            Promise.resolve({
            name: 'feature',
            enabled: true,
            strategies: [
                {
                    name: 'default',
                },
            ],
        });
    })
}
