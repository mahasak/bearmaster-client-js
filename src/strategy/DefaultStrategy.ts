import { Strategy } from './strategy';

export class DefaultStrategy extends Strategy {
    constructor() {
        super('default');
    }

    isEnabled() {
        return true;
    }
}