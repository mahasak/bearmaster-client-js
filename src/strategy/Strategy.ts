import { IExperimentStrategy, IExperimentConstraint, IExperimentContext } from "../interfaces";
import { Operator } from "../enums";

const resolveContextValue = (context: IExperimentContext, field: string) => {
    if (context[field]) {
        return context[field];
    } else if (context.properties && context.properties[field]) {
        return context.properties[field];
    } else {
        return undefined;
    }
}

export abstract class Strategy implements IExperimentStrategy {
    public name: string;
    private returnValue: boolean;

    constructor(name: string, returnValue: boolean = false) {
        this.name = name || 'unknown';
        this.returnValue = returnValue;
    }

    checkConstraint(constraint: IExperimentConstraint, context: IExperimentContext): boolean {
        const field = constraint.contextName;
        const contextValue = resolveContextValue(context, field);
        const isIn = constraint.values.some(val => val.trim() === contextValue);
        return constraint.operator === Operator.IN ? isIn : !isIn;
    }

    checkConstraints(constraints: IExperimentConstraint[], context: IExperimentContext): boolean {
        if (!constraints || constraints.length === 0) {
            return true;
        }
        return constraints.every(constraint => this.checkConstraint(constraint, context));
    }

    isEnabled(params: any, context: IExperimentContext): boolean {
        return this.returnValue;
    }

    isEnabledWithConstraint(params: any, context: IExperimentContext, constraints: IExperimentConstraint[]): boolean {
        return this.checkConstraints(constraints, context) && this.isEnabled(params, context);
    }
}
