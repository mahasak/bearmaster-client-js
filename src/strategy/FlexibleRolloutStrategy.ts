import { Strategy } from './strategy';
import { normalizedValue } from '../utils';
import { IExperimentContext } from '../interfaces';

const STICKINESS = {
    default: 'default',
    userId: 'userId',
    sessionId: 'sessionId',
    random: 'random',
};

export class FlexibleRolloutStrategy extends Strategy {
    private randomGenerator: Function = () => Math.round(Math.random() * 100) + 1 + '';

    constructor(radnomGenerator?: Function) {
        super('flexibleRollout');
        if (radnomGenerator) {
            this.randomGenerator = radnomGenerator;
        }
    }

    resolveStickiness(stickiness: string, context?: IExperimentContext): any {
        switch (stickiness) {
            case STICKINESS.userId:
                return context?.userId;
            case STICKINESS.sessionId:
                return context?.sessionId;
            case STICKINESS.random:
                return this.randomGenerator();
            default:
                return context?.userId || context?.sessionId || this.randomGenerator();
        }
    }

    isEnabled(parameters: any, context?: IExperimentContext) {
        const groupId = parameters.groupId || context?.featureToggle || '';
        const percentage = Number(parameters.rollout);
        const stickiness = parameters.stickiness || STICKINESS.default;
        const stickinessId = this.resolveStickiness(stickiness, context);

        if (!stickinessId) {
            return false;
        } else {
            const normalizedUserId = normalizedValue(stickinessId, groupId);
            return percentage > 0 && normalizedUserId <= percentage;
        }
    }
}
