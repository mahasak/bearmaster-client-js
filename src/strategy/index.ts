import { IExperimentStrategy } from "../interfaces";
import { DefaultStrategy } from "./DefaultStrategy";

export const defaultStrategies: IExperimentStrategy[] = [
    new DefaultStrategy(),
];