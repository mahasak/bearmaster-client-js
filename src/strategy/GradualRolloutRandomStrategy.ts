import { Strategy } from './Strategy';
import { IExperimentContext } from '../interfaces';

export class GradualRolloutRandomStrategy extends Strategy {
    private randomGenerator: Function = () => Math.floor(Math.random() * 100) + 1;

    constructor(randomGenerator?: Function) {
        super('gradualRolloutRandom');
        this.randomGenerator = randomGenerator || this.randomGenerator;
    }

    isEnabled(parameters: any, context?: IExperimentContext) {
        const percentage: number = Number(parameters.percentage);
        const random: number = this.randomGenerator();
        return percentage >= random;
    }
}
