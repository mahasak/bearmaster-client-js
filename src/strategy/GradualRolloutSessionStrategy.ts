import { Strategy } from './Strategy';
import { normalizedValue } from '../utils';
import { IExperimentContext } from '../interfaces';

export class GradualRolloutSessionStrategy extends Strategy {
    constructor() {
        super('gradualRolloutSessionId');
    }

    isEnabled(parameters: any, context?: IExperimentContext) {
        const sessionId = context?.sessionId;
        if (!sessionId) {
            return false;
        }

        const percentage = Number(parameters.percentage);
        const groupId = parameters.groupId || '';

        const normalizedId = normalizedValue(sessionId, groupId);

        return percentage > 0 && normalizedId <= percentage;
    }
}
