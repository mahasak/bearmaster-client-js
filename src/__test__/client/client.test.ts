import Client from '../../client';
import { Strategy } from '../../strategy/Strategy'
import {
    IExperimentVariantDefinition,
    IExperimentStrategyInfo,
    IExperiment,
    IExperimentStrategy
} from '../../interfaces';
import { setup } from '../repository/setup';
import { MockStorage } from '../mock/MockStorage';
import Repository from '../../repository';
import { DefaultStrategy } from '../../strategy/DefaultStrategy';
import { MockRepository } from '../mock/MockRepository';

const appName = 'foo';
const instanceId = 'bar';

function buildToggle(name: string, active: boolean, strategies?: IExperimentStrategyInfo[], variants: IExperimentVariantDefinition[] = []): IExperiment {
    return {
        name,
        enabled: active,
        strategies: strategies || [{ name: 'default', parameters: {}, constraints: [] }],
        variants,
    };
}

class CustomStrategy extends Strategy {
    constructor() {
        super('custom');
    }

    isEnabled() {
        return true;
    }
}

class CustomFalseStrategy extends Strategy {
    constructor() {
        super('custom-false');
    }

    isEnabled() {
        return false;
    }
}

const log = (err: any) => {
    console.error(err);
};

test('should use provided repository', () => {
    const repo = new MockRepository({
        backupPath: 'foo',
        url: 'http://experiment-server-client-01.mahasak.com',
        appName,
        instanceId,
        refreshInterval: 0
    }, () => buildToggle('feature', true))
    const client = new Client(repo, [new DefaultStrategy()]);
    client.on('error', log).on('warn', log);
    const result = client.isEnabled('feature');

    expect(result).toBeTruthy();
});

test('should fallback when missing feature', () => {
    const repo = new MockRepository({
        backupPath: 'foo',
        url: 'http://experiment-server-client-01.mahasak.com',
        appName,
        instanceId,
        refreshInterval: 0
    }, () => null)
    const client = new Client(repo, []);
    client.on('error', log).on('warn', log);

    const result = client.isEnabled('feature-x', {}, () => false);
    expect(result).toBeFalsy();

    const result2 = client.isEnabled('feature-x', {}, () => true);
    expect(result2).toBeTruthy();
});

test('should consider toggle not active', () => {
    const repo = new MockRepository({
        backupPath: 'foo',
        url: 'http://experiment-server-client-01.mahasak.com',
        appName,
        instanceId,
        refreshInterval: 0
    }, () => buildToggle('feature', false))

    const client = new Client(repo, [new DefaultStrategy()]);
    client.on('error', log).on('warn', log);
    const result = client.isEnabled('feature');

    expect(result).toBeFalsy();
});

test('should use custom strategy', () => {
    const repo = new MockRepository({
        backupPath: 'foo',
        url: 'http://experiment-server-client-01.mahasak.com',
        appName,
        instanceId,
        refreshInterval: 0
    }, () => buildToggle('feature', true, [{ name: 'custom', parameters: {}, constraints: [] }]))
    

    const client = new Client(repo, [new DefaultStrategy(), new CustomStrategy()]);
    client.on('error', log).on('warn', log);
    const result = client.isEnabled('feature');

    expect(result).toBeTruthy();
});

test('should use a set of custom strategies', () => {
    const repo = new MockRepository({
        backupPath: 'foo',
        url: 'http://experiment-server-client-01.mahasak.com',
        appName,
        instanceId,
        refreshInterval: 0
    }, () => buildToggle('feature', true, [{ name: 'custom', parameters: {}, constraints: []  }, { name: 'custom-false' , parameters: {}, constraints: [] }]))

    const strategies = [new CustomFalseStrategy(), new CustomStrategy()];
    const client = new Client(repo, strategies);
    client.on('error', log).on('warn', log);
    const result = client.isEnabled('feature');

    expect(result).toBeTruthy();
});

test('should return false a set of custom-false strategies', () => {
    const repo = new MockRepository({
        backupPath: 'foo',
        url: 'http://experiment-server-client-01.mahasak.com',
        appName,
        instanceId,
        refreshInterval: 0
    }, () => buildToggle('feature', true, [
        { name: 'custom-false', parameters: {}, constraints: [] },
        { name: 'custom-false', parameters: {}, constraints: [] },
    ]))

    const strategies = [new CustomFalseStrategy(), new CustomStrategy()];
    const client = new Client(repo, strategies);
    client.on('error', log).on('warn', log);
    const result = client.isEnabled('feature');

    expect(result === false).toBeTruthy;
});

test('should emit error when invalid feature runtime', () => {
    const repo = new MockRepository({
        backupPath: 'foo',
        url: 'http://experiment-server-client-01.mahasak.com',
        appName,
        instanceId,
        refreshInterval: 0
    }, () => { return {
        name: 'feature-malformed-strategies',
        enabled: true,
        strategies: true,
    }})

    const strategies: IExperimentStrategy[] = [new DefaultStrategy()];
    const client = new Client(repo, strategies);
    client.on('error', (err) => {
        expect(err).toBeTruthy();
        expect(err.message).toContain('Malformed feature');
    })
    
    expect(client.isEnabled('feature-malformed-strategies')).toBeFalsy();
    expect
});
