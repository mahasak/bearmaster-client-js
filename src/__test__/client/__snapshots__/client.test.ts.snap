// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return variant when equal weight on 1,1,1 1`] = `
Object {
  "enabled": true,
  "name": "variant1",
  "params": Array [
    Object {
      "name": "p1",
      "type": "string",
      "value": "val1",
    },
  ],
}
`;

exports[`should return variant when equal weight on 1,1,1 2`] = `
Object {
  "enabled": true,
  "name": "variant2",
  "params": Array [
    Object {
      "name": "p1",
      "type": "string",
      "value": "val2",
    },
  ],
}
`;

exports[`should return variant when equal weight on 1,1,1 3`] = `
Object {
  "enabled": true,
  "name": "variant3",
  "params": Array [
    Object {
      "name": "p1",
      "type": "string",
      "value": "val3",
    },
  ],
}
`;

exports[`should return variant when equal weight on 33,33,33 1`] = `
Object {
  "enabled": true,
  "name": "variant1",
  "params": Array [
    Object {
      "name": "p1",
      "type": "string",
      "value": "val1",
    },
  ],
}
`;

exports[`should return variant when equal weight on 33,33,33 2`] = `
Object {
  "enabled": true,
  "name": "variant2",
  "params": Array [
    Object {
      "name": "p1",
      "type": "string",
      "value": "val2",
    },
  ],
}
`;

exports[`should return variant when equal weight on 33,33,33 3`] = `
Object {
  "enabled": true,
  "name": "variant3",
  "params": Array [
    Object {
      "name": "p1",
      "type": "string",
      "value": "val3",
    },
  ],
}
`;

exports[`should return variant when equal weight on 100,100,100 1`] = `
Object {
  "enabled": true,
  "name": "variant1",
  "params": Array [
    Object {
      "name": "p1",
      "type": "string",
      "value": "val1",
    },
  ],
}
`;

exports[`should return variant when equal weight on 100,100,100 2`] = `
Object {
  "enabled": true,
  "name": "variant2",
  "params": Array [
    Object {
      "name": "p1",
      "type": "string",
      "value": "val2",
    },
  ],
}
`;

exports[`should return variant when equal weight on 100,100,100 3`] = `
Object {
  "enabled": true,
  "name": "variant3",
  "params": Array [
    Object {
      "name": "p1",
      "type": "string",
      "value": "val3",
    },
  ],
}
`;
